crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -805,62 +807,18 @@ where
         }
     }
 
-    fn visit_parameters(&mut self, parameters: &'ast ruff_python_ast::Parameters) {
+    fn visit_parameters(&mut self, parameters: &'ast ast::Parameters) {
         // Intentionally avoid walking default expressions, as we handle them in the enclosing
         // scope.
         for parameter in parameters.iter().map(ast::AnyParameterRef::as_parameter) {
             self.visit_parameter(parameter);
         }
     }
 
-    fn visit_pattern(&mut self, pattern: &'ast ast::Pattern) {
-        // The definition visitor will recurse into the pattern so avoid walking it here.
-        let mut definition_visitor = MatchPatternDefinitionVisitor::new(self, pattern);
-        definition_visitor.visit_pattern(pattern);
-    }
-}
-
-/// A visitor that adds symbols and definitions for the identifiers in a match pattern.
-struct MatchPatternDefinitionVisitor<'a, 'db> {
-    /// The semantic index builder in which to add the symbols and definitions.
-    builder: &'a mut SemanticIndexBuilder<'db>,
-    /// The index of the current node in the pattern.
-    index: u32,
-    /// The pattern being visited. This pattern is the outermost pattern that is being visited
-    /// and is required to add the definitions.
-    pattern: &'a ast::Pattern,
-}
-
-impl<'a, 'db> MatchPatternDefinitionVisitor<'a, 'db> {
-    fn new(builder: &'a mut SemanticIndexBuilder<'db>, pattern: &'a ast::Pattern) -> Self {
-        Self {
-            index: 0,
-            builder,
-            pattern,
-        }
-    }
-
-    fn add_symbol_and_definition(&mut self, identifier: &ast::Identifier) {
-        let symbol = self
-            .builder
-            .add_or_update_symbol(identifier.id().clone(), SymbolFlags::IS_DEFINED);
-        self.builder.add_definition(
-            symbol,
-            MatchPatternDefinitionNodeRef {
-                pattern: self.pattern,
-                identifier,
-                index: self.index,
-            },
-        );
-    }
-}
-
-impl<'ast, 'db> Visitor<'ast> for MatchPatternDefinitionVisitor<'_, 'db>
-where
-    'ast: 'db,
-{
-    fn visit_expr(&mut self, expr: &'ast ast::Expr) {
-        self.builder.visit_expr(expr);
+    fn visit_match_case(&mut self, match_case: &'ast ast::MatchCase) {
+        self.pattern_state = Some(MatchPatternState::new(&match_case.pattern));
+        walk_match_case(self, match_case);
+        self.pattern_state = None;
     }
```

Comment:
I think I might want to limit the `pattern_state` to be `Some` only during the `pattern` visit. Currently, it's `Some` during the entire `MatchCase` visit.